src/typecheck/float.ml

```diff
@@ -0,0 +1,240 @@
+        | _ -> assert false (* Because to_fp is specified below with 2 parameters *)
+      in
+      let args = List.map (Type.parse_term env) args in
+      let return t = Some (Type.Term t) in
+      match args with
+      | [ a ] -> return (T.ieee_format_to_fp e s a)
+      | [ rm; b ] -> begin
+          match T.type_for_to_fp b with
+          | `Real -> return (T.real_to_fp e s rm b)
+          | `Bitv -> return (T.ubv_to_fp e s rm b)
+          | `Float -> return (T.fp_to_fp e s rm b)
+          | `Other ->
+            let err = To_fp_incorrect_args in
+            raise (Type.Typing_error (err, env, ast))
+        end
+      | _ ->
+        let err = To_fp_incorrect_args in
+        raise (Type.Typing_error (err, env, ast))
+
+    let parse _version env ast s args =
+      match s with
+      (* sort *)
+      | Type.Id { Id.ns = Id.Sort; name; } ->
+        parse_id env ast name [
+          "FloatingPoint", 2, (function
+              | [n_e;n_s] ->
+                fp_type env ast args (parse_int env ast n_e) (parse_int env ast n_s)
+              | _ -> assert false);
+        ] (function
+            | ["Float16"] -> fp_type env ast args 5 11
+            | ["Float32"] -> fp_type env ast args 8 24
+            | ["Float64"] -> fp_type env ast args 11 53
+            | ["Float128"] -> fp_type env ast args 15 113
+            | ["RoundingMode"] ->
+              Base.make_op0 (module Type) env ast "RoundingMode" args
+                (fun () -> Type.Ty Ty.roundingMode)
+            | _ -> None)
+      (* terms *)
+      | Type.Id { Id.ns = Id.Term; name; } ->
+        parse_id env ast name [
+          "+oo", 2, app0_param2 env ast args "plus_infinity" T.plus_infinity;
+          "-oo", 2, app0_param2 env ast args "minus_infinity" T.minus_infinity;
+          "+zero", 2, app0_param2 env ast args "plus_zero" T.plus_zero;
+          "-zero", 2, app0_param2 env ast args "minus_zero" T.minus_zero;
+          "NaN", 2, app0_param2 env ast args "nan" T.nan;
+          "to_fp", 2, to_fp env ast args;
+          "to_fp_unsigned", 2, app2_param2 env ast args "ubv_to_fp" T.ubv_to_fp;
+          "fp.to_ubv", 1, app2_param1 env ast args "to_ubv" T.to_ubv;
+          "fp.to_sbv", 1, app2_param1 env ast args "to_sbv" T.to_sbv;
+        ] (function
+            | ["fp"] ->
+              let parse_bitv_lit arg =
+                match arg.Term.term with
+                | Term.Symbol { Id.ns = Id.Value Id.Binary; name; } ->
+                  parse_binary env ast name
+                | Term.Symbol { Id.ns = Id.Value Id.Hexadecimal; name; } ->
+                  parse_hexa env ast name
+                | _ ->
+                  let err = Bitvector_litteral_expected in
+                  raise (Type.Typing_error (err, env, ast))
+              in
+              Base.make_op3 (module Type) env ast "fp" args
+                (fun (a,b,c) ->
+                   Type.Term (T.fp (parse_bitv_lit a) (parse_bitv_lit b) (parse_bitv_lit c)))
```

Comment:
Here in the case of dolmen, just parsing the arguments of `fp` using the functions provided in `Type` would be fine, as assuming the bitvector theory is also in the builtins (see later), the arguments would be parsed and type-checked as bitvector terms, which you could then use to build the application of `fp`to these bitvectors.

Doing things this way would actually lead to the question of how to express dependencies between theories (in this case, the Float theory would depend on the Bitvector one). It would be quite involved to set-up a general purpose mechanism for theories to be named and declare dependencies I think, so we're left with a few best-effort solutions:
- document the dependency and trust users to respect it, but that's a bit brittle
- actually include the bitvector theory in the float theory e.g. by:
  - taking the theory as additional functor argument and calling its parse function in the parse function of the float theory
  - declaring the required interface for the float functor to also include the requirement for the bitvector one, instantiate the bitvector theory inside the functor for floats, and add it to the parse function
- in this case, considering what's actually needed is only a small part of the bitvector theory, we have ad-hoc (but probably very good) solutions:
  - the one currently adopted which copies the code for building bitvector litterals
  - this code could actually be aliased to the bitvector code using one of the two solutions above (probably rather by taking the bitv theory as functor argument, since duplicating the bitv theory functor instantiation just for a small part of its code would be a bit overkill).

I think ad-hoc solutions are okay in this case because it shouldn't happen often, since theories for type-checking often match the semantics theories used by SMT solvers, which are often required to not share symbols with each other.

Reply:
I agree that if the smtlib specifies that it is only for literals, then dolmen should strictly parse and type exactly what is specified indeed.

Reply:
In this case (i.e. assuming it is for literals only), the current implementation, is well, not yet optimal, primarily because in case of an error on the size of the first argument, the typed term corresponding to the first string is never really seen by the core type-checker (nor by the float theory actually), which will present problems for accurate error reporting, but it needs a bit of context:

Currently, when typing a term such as `f x y z ...`, if one of the arguments has an incorrect type w.r.t. the type(s) of the function `f`, the error reported will concern the whole application, which makes its location a bit imprecise. Instead, dolmen should report the location of the argument whose type is wrong. This is a bit tricky, but basically, all of the potential solutions rely on the fact that we can associate a location to each typed terms (either in an association list created for each application, or through a global table), and for that, the type-checker need to be aware of most of the typed terms created.

This conflicts with the fact that the typed term in the typing error when the first string is not of length 1, cannot be associated to a location. That would be very naturally solved if `fp` was a regular function, at least in `Dolmen.Expr.Term`. The float theory could still enforce that it is applied to bitvector literals, but it could try and build the typed bitvectors before trying to build the application ?

Reply:
I'll change `fp` to take any bitvector as arguments (thus making the bitv theory a dependency of the fp theory in practice). We'll change that if needed.